onadata/apps/api/viewsets/xform_viewset.py

```diff
@@ -423,24 +421,22 @@ def enketo(self, request, **kwargs):
             request_vars = request.GET
             defaults = generate_enketo_form_defaults(
                 self.object, **request_vars)
-            url = enketo_url(
+            data = enketo_url(
                 form_url, self.object.id_string, **defaults)
-            preview_url = get_enketo_preview_url(request,
-                                                 self.object.user.username,
-                                                 self.object.id_string,
-                                                 xform_pk=self.object.pk)
+            offline_url = (data.get('edit_url') or data.get('offline_url') or
+                           data.get('url'))
```

Comment:
A couple of questions:
- Why is `edit_url` given more priority than `offline_url` when assigning to `offline_url` variable
- Why are we using `or` here? Do `edit_url`,  `offline_url` and `url` all have the same values?

Reply:
> Do edit_url, offline_url and url all have the same values?

Yes.

> Why is edit_url given more priority than offline_url when assigning to offline_url variable

No priority given. Changed the order

Reply:
> Do edit_url, offline_url and url all have the same values? Yes

Do you mind printing out here the output of `data` dictionary?

Reply:
This is the sample output of this request
`http -a <enketo-key>: POST :8005/api_v2/survey/all server_url=="https://api.ona.io/<username>" form_id=<id_string>`

```json
{
    "code": 200,
    "enketo_id": "oTd4kC2o",
    "iframe_url": "http://localhost:8005/i/::oTd4kC2o",
    "offline_url": "http://localhost:8005/x/#oTd4kC2o",
    "preview_iframe_url": "http://localhost:8005/preview/i/::oTd4kC2o",
    "preview_url": "http://localhost:8005/preview/::oTd4kC2o",
    "single_iframe_url": "http://localhost:8005/single/i/::oTd4kC2o",
    "single_once_iframe_url": "http://localhost:8005/single/i/::85e2219727661e69ecc79345bbd299c6",
    "single_once_url": "http://localhost:8005/single/::85e2219727661e69ecc79345bbd299c6",
    "single_url": "http://localhost:8005/single/::oTd4kC2o",
    "url": "http://localhost:8005/::oTd4kC2o"
}
```

Comparing the value of 
```json
"offline_url": "http://localhost:8005/x/#oTd4kC2o",
"url": "http://localhost:8005/::oTd4kC2o"
```

Maybe just the /x/?
Or should I pick out specifically the offline url here?

Reply:
Cool, so a couple of things:
- `offline_url` and `url` are different; `offline_url`, as the name suggests, can be used when you don't have internet connection to collect data while `url` cannot be used
- I can't seem to see `edit_url` from the payload. Where are we adding it from and may be, why should we add it?

Reply:
I guess what i'm trying to say is that we only need the `offline_url`